pkg/dclient/client.go

```diff
@@ -24,18 +20,47 @@ import (
 	"k8s.io/client-go/rest"
 )
 
-//Client enables interaction with k8 resource
-type Client struct {
+type Interface interface {
+	// NewDynamicSharedInformerFactory returns a new instance of DynamicSharedInformerFactory
+	NewDynamicSharedInformerFactory(time.Duration) dynamicinformer.DynamicSharedInformerFactory
+	// GetEventsInterface provides typed interface for events
+	GetEventsInterface() (event.EventInterface, error)
+	// GetCSRInterface provides type interface for CSR
+	GetCSRInterface() (csrtype.CertificateSigningRequestInterface, error)
+	// GetDynamicInterface fetches underlying dynamic interface
+	GetDynamicInterface() dynamic.Interface
+	// Discovery return the discovery client implementation
+	Discovery() IDiscovery
+	// SetDiscovery sets the discovery client implementation
+	SetDiscovery(discoveryClient IDiscovery)
+	// GetResource returns the resource in unstructured/json format
+	GetResource(apiVersion string, kind string, namespace string, name string, subresources ...string) (*unstructured.Unstructured, error)
+	// PatchResource patches the resource
+	PatchResource(apiVersion string, kind string, namespace string, name string, patch []byte) (*unstructured.Unstructured, error)
+	// ListResource returns the list of resources in unstructured/json format
+	// Access items using []Items
+	ListResource(apiVersion string, kind string, namespace string, lselector *meta.LabelSelector) (*unstructured.UnstructuredList, error)
+	// DeleteResource deletes the specified resource
+	DeleteResource(apiVersion string, kind string, namespace string, name string, dryRun bool) error
+	// CreateResource creates object for the specified resource/namespace
+	CreateResource(apiVersion string, kind string, namespace string, obj interface{}, dryRun bool) (*unstructured.Unstructured, error)
+	// UpdateResource updates object for the specified resource/namespace
+	UpdateResource(apiVersion string, kind string, namespace string, obj interface{}, dryRun bool) (*unstructured.Unstructured, error)
+	// UpdateStatusResource updates the resource "status" subresource
+	UpdateStatusResource(apiVersion string, kind string, namespace string, obj interface{}, dryRun bool) (*unstructured.Unstructured, error)
+}
+
+// Client enables interaction with k8 resource
+type client struct {
 	client          dynamic.Interface
 	log             logr.Logger
 	clientConfig    *rest.Config
 	kclient         kubernetes.Interface
 	DiscoveryClient IDiscovery
```

Comment:
Shall we make this private now that we have getters and setters?

Reply:
Good catch, fixed.